Add notifications for volunteer birthdays (#5316)

davidgumberg opened the pull request:
Resolves #5316

This PR adds notifications for volunteer birthdays.

- Adds an optional 'date_of_birth' parameter to the Users model, which is a prerequisite for being able to notify supervisors of upcoming volunteer birthdays.

- Adds a `birthday_next_month` scope to `Volunteer`, volunteer birthday notification, and rake task for creating the notifications.

## Permissions
Volunteers can edit their own DOB, and supervisors and admins from the same org as a volunteer are able to edit the volunteer's DOB.

## Tests
I've added tests for the views, and permissions, and a validation rule for the DOB's (they must be in the past.) and for the notification message, and notification rake task.

![Volunteer birthday notification](https://github.com/rubyforgood/casa/assets/2257631/632659cc-746d-4c30-8eba-4c6d18afdb5a)

![Supervisor editing a volunteer from the shared form](https://github.com/rubyforgood/casa/assets/2257631/b5d06cb0-d32a-4962-84cd-415cdb8cc98b)

![User editing their own profile with datepicker](https://github.com/rubyforgood/casa/assets/2257631/23e00ae6-2888-4ade-99dc-7cf69b33df0a)

![Date of birth validation message](https://github.com/rubyforgood/casa/assets/2257631/05799a22-a87e-4819-8ac2-a456fa80755e)

davidgumberg committed:
Add tests for users 'date_of_birth' field

mattzollinhofer commented on app/validators/user_validator.rb:
```diff
@@ -36,4 +37,16 @@ def valid_phone_number_if_receive_sms_notifications(record)
     end
   end
+
+  def valid_date_of_birth(date_of_birth, record)
+    return true unless date_of_birth.present?
+
+    # Date of birth must be in the past
+    if date_of_birth.to_time >= Time.zone.now
+      record.errors.add(:base, " Date of birth must be in the past.")
+      false
+    else
+      true
+    end
+  end
```
Again, I'm new here so take with a giant grain of salt.

I don't believe custom validators need to return true or false. They only need to add an error to the record or not. with that said I here's a suggested refactoring.

```ruby
  def valid_date_of_birth(date_of_birth, record)
    return unless date_of_birth.present?

    record.errors.add(:base, "Date of birth must be in the past.") if date_of_birth.past?
                                                     #                              ^^^^^
                                                     # I believe this can do the check for in the past.
  end
```

### References
validators ref: https://guides.rubyonrails.org/active_record_validations.html#performing-custom-validations
`past?` ref: https://api.rubyonrails.org/v7.0/classes/DateAndTime/Calculations.html#method-i-past-3F

davidgumberg replied:
Good catch, thanks

mattzollinhofer reviewed (commented):
This is a completely drive by review from a new guy to the area. Please take it or leave it completely. Just wanted to provide a few thoughts. Generally looks good!

schoork commented:
@davidgumberg What's the reasoning behind splitting this into 2 prs?

davidgumberg commented on app/models/learning_hour.rb:
```diff
@@ -36,6 +36,7 @@ def user_org_learning_topic_enable?
 #  id                     :bigint           not null, primary key
 #  duration_hours         :integer          not null
 #  duration_minutes       :integer          not null
+#  learning_type          :integer          default(5)
```
This is some lint from the annotate gem, and this annotation is correct, so I figured I would leave it in, splitting out into a different commit is a good idea.

davidgumberg committed:
'annotate' gem lint

davidgumberg committed:
Add 'date of birth' to Users model

davidgumberg commented:
> @davidgumberg What's the reasoning behind splitting this into 2 prs?

More than the size of the changes, I just wanted to make sure that adding a date of birth column to the Users table was desired, but since it's only use case is the birthday notification I think I'll make it one PR so they can be looked at together.

davidgumberg committed:
Add short ordinal date format

davidgumberg committed:
Add volunteer birthday notification and task

davidgumberg committed:
Add tests for volunteer birthday notifications

schoork commented on app/views/users/edit.html.erb:
```diff
@@ -23,6 +23,14 @@
+            <div class="input-style-1">
+              <%= form.label :date_of_birth, "Date of birth" %>
+              <%= form.text_field :date_of_birth,
+                                  value: @user.date_of_birth&.strftime("%Y/%m/%d"),
```
Less concerned about this `date_of_birth&.strftime...` being in the decorator, but I'm not a huge fan of the safe operator in views.

davidgumberg replied:
The decorator is a good place to handle nil DOB more explicitly anyways 👍

schoork commented on app/views/volunteers/_form.html.erb:
```diff
@@ -16,6 +16,14 @@
+  <div class="input-style-1">
+    <%= form.label :date_of_birth, "Date of birth" %>
+    <%= form.text_field :date_of_birth,
+                        value: @volunteer.date_of_birth&.strftime("%Y/%m/%d"),
```
Same as app/views/users/edit.html.erb

schoork commented on app/views/volunteers/_form.html.erb:
```diff
@@ -16,6 +16,14 @@
+    <%= form.label :date_of_birth, "Date of birth" %>
+    <%= form.text_field :date_of_birth,
+                        value: @volunteer.date_of_birth&.strftime("%Y/%m/%d"),
+                        data: {provide: "datepicker", date_format: "yyyy/mm/dd"},
```
See comment for config/initializers/date_formats.rb

schoork commented on config/initializers/date_formats.rb:
This is really cool! Can we leverage in some other places? Like say on app/views/volunteers/_form.html.erb for the format for the date picker?

schoork commented on spec/lib/tasks/volunteer_birthday_reminder_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    Rake::Task.clear
+    Casa::Application.load_tasks
+
+    travel_to Date.new(2022, 10, 1)
```
So cool!

schoork commented:
Absolutely love all the tests @davidgumberg!

javiguerra777 commented on app/controllers/casa_admins_controller.rb:
```diff
@@ -135,7 +135,7 @@ def set_admin
 
   def update_casa_admin_params
-    CasaAdminParameters.new(params).with_only(:email, :display_name, :phone_number, :monthly_learning_hours_report)
+    CasaAdminParameters.new(params).with_only(:email, :display_name, :phone_number, :date_of_birth, :monthly_learning_hours_report)
```
Good job adding date of birth to the accepted params to meet the requirements of PR.

davidgumberg committed:
Move user DOB formats to decorator

davidgumberg committed:
Add spec for DOB formatters

davidgumberg committed:
Add a helper for the notifications that are marked read once seen

davidgumberg commented on app/models/volunteer.rb:
```diff
@@ -57,6 +63,10 @@ class Volunteer < User
                                  }
 
+  scope :birthday_next_month, -> {
+    where("EXTRACT(month from date_of_birth) = ?", DateTime.now.next_month.month)
```
I tried, but could not come up with a way to do this without using EXTRACT

schoork replied:
Try this
```
where(date_of_birth: Date.today.beginning_of_month.next_month..Date.today.end_of_month.next_month)
```

davidgumberg replied:
The problem is that this range will only include date_of_birth's that fall on the next month of this year.

schoork replied:
Good point. Yeah, the only solution I saw started to feel over-engineered. There's no user input in this, so that makes me more comfortable. Keep as is.

davidgumberg commented:
@schoork Thanks for the review, I've implemented the suggestions you made (minus using rails helpers for `birthday_next_month`, which I could not figure out a way to do) and I've also added tests for the two new scopes in the volunteer model.

schoork commented on app/decorators/user_decorator.rb:
```diff
@@ -44,4 +44,16 @@ def formatted_reset_password_sent_at
   end
+
+  def formatted_birthday
+    return "" unless object.date_of_birth.respond_to?(:strftime)
```
Love the NoMethodError guards.

schoork reviewed (approved):
Really awesome work @davidgumberg! Love all the tests. Had never heard of the `travel_to DATE`. That's pretty nifty!

davidgumberg commented:
> Really awesome work @davidgumberg! Love all the tests. Had never heard of the `travel_to DATE`. That's pretty nifty!

Thanks for the review and feedback!

I got the idea for `travel_to` from [`spec/lib/tasks/emancipation_checklist_reminder_spec.rb` ](https://github.com/rubyforgood/casa/blame/e96ff2a5d637a9f90a10f8abb10ea1c8e85818d8/spec/lib/tasks/emancipation_checklist_reminder_spec.rb#L12)

bcastillo32 commented:
@davidgumberg Thanks for the work here. I want to share this update with the casa stakeholders. They will love it. how many days out does the birthday need to be to receive a notification?

davidgumberg commented:
> @davidgumberg Thanks for the work here. I want to share this update with the casa stakeholders. They will love it. how many days out does the birthday need to be to receive a notification?

Thanks @bcastillo32.

Actually, there is a heroku task that will need to be added to run once a month by whoever administers the CASA application, and it will notify volunteers in the same cadence as the youth volunteer reminders, on the 15th or so of each month, they will receive a notification about all birthdays that are coming in the next month. Please let me know if that behavior is aligned with what the stakeholders requested, and if not, I can write another PR to update the changes to better reflect their request.

bcastillo32 commented:
> > @davidgumberg Thanks for the work here. I want to share this update with the casa stakeholders. They will love it. how many days out does the birthday need to be to receive a notification?
> 
> Thanks @bcastillo32.
> 
> Actually, there is a heroku task that will need to be added to run once a month by whoever administers the CASA application, and it will notify volunteers in the same cadence as the youth volunteer reminders, on the 15th (or whichever day of the month the administrator sets the task to run) of each month, users will receive a notification about all volunteer birthdays that are coming in the next month. Please let me know if that behavior is aligned with what the stakeholders requested, and if not, I can write another PR to update the changes to better reflect their request.

thanks @davidgumberg ! I will reach out to the CASA team to confirm if the proposed schedule suits their requirements. I plan to inform them of the suggested cadence, and if they have any objections and want to propose a different one, we can consider it. Thank you once again for your valuable input!
